### PR TITLE
feat: vllm openai api integration

### PR DESCRIPTION
**Pull request by jorgeantonio21**

### CODE REVIEW

**Review by Cifko — APPROVED** [review]

File: backends/vllm/src/llm_engine.rs [hunk 1]
```diff
@@ -380,10 +393,12 @@ impl LlmEngine {
             };
 
             sequence_guard_lock.output_text.push_str(&generated_token);
+            // TODO: send generated token to the client, if streaming is enabled
```
[the remark this answers is not in the document]
Comment (jorgeantonio21) [reply backends/vllm/src/llm_engine.rs:396]: Yes there is. Streaming will be added on a different PR, once this one is merged.

File: server/src/api/chat_completions.rs [hunk 2]
```diff
@@ -328,20 +514,171 @@ impl RequestBody {
...
+            .iter()
+            .map(|o| o.token_ids.len())
+            .sum::<usize>() as u32;
+        let total_tokens = prompt_tokens + completion_tokens;
+
+        let now = Instant::now();
+        let finished_time = value
+            .metrics
+            .read()
+            .unwrap()
```
Comment (eureka-cpu) [inline server/src/api/chat_completions.rs:644]: maybe use an expect here instead of unwrap

File: server/src/api/chat_completions.rs [hunk 3]
```diff
@@ -328,20 +514,171 @@ impl RequestBody {
...
+            .finished_time
+            .ok_or("Finished time not found")?;
+        let duration = now.duration_since(finished_time);
+        let system_now = SystemTime::now();
+        let system_duration = system_now.duration_since(SystemTime::UNIX_EPOCH).unwrap();
```
Comment (eureka-cpu) [inline server/src/api/chat_completions.rs:649]: this could also be an expect

**Review by Cifko — APPROVED** [review]

**Comment by jorgeantonio21** [comment]
Closes #77.

File: backends/vllm/src/llm_engine.rs [hunk 4]
```diff
@@ -152,13 +157,21 @@ impl LlmEngine {
...
                 //    we do so to optimize GPU utilization. This is
                 //    supposed to be safe
                 if !request_outputs.is_empty() {
-                    self.atoma_client_sender.send(request_outputs)?;
+                    // 4. Extract the response sender from the `response_senders` hashmap
+                    for request_output in request_outputs {
+                        let response_sender = self
+                            .response_senders
+                            .remove(&request_output.request_id)
+                            .unwrap();
```
Comment (francis2tm) [inline backends/vllm/src/llm_engine.rs:170]: unwrap?

File: backends/vllm/src/config.rs [hunk 5]
```diff
@@ -453,6 +453,36 @@ pub enum SchedulerConfigError {
...
+
+impl ValidationConfig {
+    /// Creates a new instance of `ValidationConfig` from a `.toml` file.
+    pub fn from_file_path<P: AsRef<Path>>(config_file_path: P) -> Self {
+        let builder = Config::builder().add_source(config::File::with_name(
+            config_file_path.as_ref().to_str().unwrap(),
+        ));
+        let config = builder
+            .build()
+            .expect("Failed to generate inference configuration file");
```
Comment (francis2tm) [inline backends/vllm/src/config.rs:479]: expect?

File: backends/vllm/src/config.rs [hunk 6]
```diff
@@ -453,6 +453,36 @@ pub enum SchedulerConfigError {
...
+        config
+            .get::<Self>("validation")
+            .expect("Failed to generated config file")
```
Comment (francis2tm) [inline backends/vllm/src/config.rs:482]: expect?

File: server/src/main.rs [hunk 7]
```diff
@@ -25,33 +45,148 @@ pub const DEFAULT_SERVER_ADDRESS: &str = "0.0.0.0";
...
 pub async fn completion_handler(
+    app_state: State<AppState>,
     headers: HeaderMap,
     Json(request): Json<RequestBody>,
-) -> impl IntoResponse {
+) -> Result<Json<ChatCompletionResponse>, (StatusCode, Json<serde_json::Value>)> {
+    let request_number = app_state.request_counter.fetch_add(1, Ordering::SeqCst);
+    let now = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Time went backwards")
```
Comment (francis2tm) [inline server/src/main.rs:186]: expect?

File: backends/vllm/src/llm_engine.rs (hunk 4 above)
Comment (jorgeantonio21) [reply backends/vllm/src/llm_engine.rs:170]: That's guaranteed to not panic.

File: backends/vllm/src/config.rs (hunk 5 above)
Comment (jorgeantonio21) [reply backends/vllm/src/config.rs:479]: Yes, the configuration is loaded at the start. If it is malformed the system can't properly be spawned, which means it is best to hard panic immediately.

File: backends/vllm/src/config.rs (hunk 6 above)
Comment (jorgeantonio21) [reply backends/vllm/src/config.rs:482]: See my response to the previous comment.

**Pull request merged**